File: CK3toEU4/Source/CK3World/Titles/DynamicTemplate.cpp
#include "DynamicTemplate.h"
#include "ParserHelpers.h"

CK3::DynamicTemplate::DynamicTemplate(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void CK3::DynamicTemplate::registerKeys()
{
	registerKeyword("key", [this](const std::string& unused, std::istream& theStream) {
		dynamicKey = commonItems::singleString(theStream).getString();
	});
	registerKeyword("tier", [this](const std::string& unused, std::istream& theStream) {
		dynamicRank = commonItems::singleString(theStream).getString();
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: CK3ToEU4/Source/CK3/Titles/DynamicTemplate.cs
using commonItems;

namespace CK3ToEU4.CK3.Titles;

class DynamicTemplate
{
	public DynamicTemplate(BufferedReader reader)
	{
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);
	}

	public string DynamicTitleKey { get; private set; } = string.Empty;
	public string DynamicTitleRank { get; private set; } = string.Empty;
		
	private void RegisterKeys(Parser parser)
	{
		parser.RegisterKeyword("key", reader => {
			DynamicTitleKey = reader.GetString();
		});
		parser.RegisterKeyword("tier", reader => {
			DynamicTitleRank = reader.GetString();
		});
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
	}
}

